src/microbe_stage/SpawnSystem.cs

```diff
@@ -253,148 +245,117 @@ private int HandleQueuedSpawns(int spawnsLeftThisFrame)
-                    float squaredDistance = displacement.LengthSquared();
-
-                    if (squaredDistance <= spawnType.SpawnRadiusSquared &&
-                        squaredDistance >= spawnType.MinSpawnRadiusSquared)
-                    {
-                        // Second condition passed. Spawn the entity.
-                        if (SpawnWithSpawner(spawnType, playerPosition + displacement, playerPosition, existing,
-                                ref spawnsLeftThisFrame, ref spawned))
-                        {
-                            estimateEntityCountInSpawnRadius += spawned;
-
-                            return;
-                        }
-                    }
-                }
+                spawns += SpawnWithSpawner(spawnType,
+                    playerLocation + new Vector3(Mathf.Cos(angle) * Constants.SPAWN_SECTOR_SIZE * 2, 0,
+                        Mathf.Sin(angle) * Constants.SPAWN_SECTOR_SIZE * 2));
             }
         }
 
-        estimateEntityCountInSpawnRadius += spawned;
-
         var metrics = PerformanceMetrics.Instance;
 
         if (metrics.Visible)
-            metrics.ReportSpawns(spawned);
+            metrics.ReportSpawns(spawns);
     }
 
     /// <summary>
     ///   Does a single spawn with a spawner
     /// </summary>
-    /// <returns>True if we have exceeded the spawn limit and no further spawns should be done this frame</returns>
-    private bool SpawnWithSpawner(Spawner spawnType, Vector3 location, Vector3 playerPosition, int existing,
-        ref int spawnsLeftThisFrame, ref int spawned)
+    private int SpawnWithSpawner(Spawner spawnType, Vector3 location)
     {
-        var enumerable = spawnType.Spawn(worldRoot, location, playerPosition);
-
-        if (enumerable == null)
-            return false;
-
-        var spawner = enumerable.GetEnumerator();
+        var spawns = 0;
 
-        while (spawner.MoveNext())
+        if (random.NextFloat() > spawnType.Density)
         {
-            if (spawner.Current == null)
-                throw new NullReferenceException("spawn enumerator is not allowed to return null");
+            return spawns;
+        }
 
-            // Spawned something
-            ProcessSpawnedEntity(spawner.Current, spawnType);
-            spawned += 1;
-            --spawnsLeftThisFrame;
+        if (spawnType is CompoundCloudSpawner || estimateEntityCount < Constants.DEFAULT_MAX_SPAWNED_ENTITIES)
+        {
+            var enumerable = spawnType.Spawn(worldRoot, location);
 
-            // Check if we are out of quota for this frame
+            if (enumerable == null)
+                return spawns;
 
-            // TODO: this is a bit awkward if this stops compound clouds from spawning as well...
-            if (spawned + existing >= Constants.DEFAULT_MAX_SPAWNED_ENTITIES)
+            using var spawner = enumerable.GetEnumerator();
+            while (spawner.MoveNext())
```

Comment:
There seems to be a massive bug here, this no longer limits the spawns per frame properly here, at least I can't see this loop exiting if there are no spawns left. Was the lag spikes the reason why you removed the colony spawns or? If so then, here's the bug which started to cause the lag spikes.

Also now that you've broken the queued spawns system, the left in code to spawn queued things, is completely useless... so I very strongly say you should fix the queued spawns system to ensure no minor stuttering happens from spawning more than a couple of things per frame.

Reply:
I fail to understand why you think that using the spawn queue would impact the sectors. The logic should be:
- If there are queued spawns, handle those
- If out of spawns this frame, skip
- Otherwise enter into the spawn sectors code. When spawning there, spawn up to spawnsLeftThisFrame things and if there are more stuff to spawn, then save that as the queue



It's simply not doable to spawn more things than spawnsLeftThisFrame, it's a massive source of small lag spikes, which I don't want to put back into the game. If you have changed the logic so that multiple spawners are allowed to trigger per frame, you need to also change the spawn queueing system to allow multiple spawners to have active queues at once, then no spawns that are wanted will be forgotten.

Reply:
Unless I'm deeply misunderstanding something (and if so educate me now) queuedSpawns is not a list of QueuedSpawn, it's a single nullable instance. That means the queue has a max length of 1. That's not enough if multiple things can spawn per frame.

Reply:
It is a nullable *enumerator* and an enumerator can return as many items as it wants. If you do `foreach(var thing in list)` that uses an enumerator. The list is converted into an enumerator and then the foreach loops it. So basically we have a nullable reference to a thing that promises to generate us one or more items we can read one at a time.

Reply:
I see how that works within a single spawntype (spawning a colony), but this code: https://github.com/Revolutionary-Games/Thrive/blob/master/src/microbe_stage/SpawnSystem.cs#L391 is not appending anything, it's replacing whatever was in the variable. This would be very hard to test, but I'm pretty sure that during a lag spike that will remove the queues from any spawn types but the last one to be run.

Reply:
I'm 99% certain that that code cannot run if the previous spawner still has stuff. Look at like 193 in Process. And unless there's a bug in HandleQueuedSpawns, it won't report that there are any spawns left if it didn't manage to spawn the entire queue.